Serialize cluster into yaml

pjanevskiTT opened the pull request:
### Issue

Solving part of #431 

### Description

Add a function to serialize cluster to the yaml (reverse of what we were doing so far). Add flag to all Cluster constructors to create mock chips inside Cluster if user wants that. Expose this (de)serializing through the API as mentioned in the issue above. Output yaml look like (note the styling difference than original, but in YAML world it is the same)

```
arch:
  1: wormhole_b0
  0: wormhole_b0
ethernet_connections:
  -
    - chip: 1
      chan: 1
    - chip: 0
      chan: 9
  -
    - chip: 1
      chan: 0
    - chip: 0
      chan: 8
chips_with_mmio:
  - 0: 0
harvesting:
  1:
    noc_translation: true
    harvest_mask: 513
  0:
    noc_translation: true
    harvest_mask: 129
boardtype:
  1: n300
  0: n300 
```

### List of the changes

- Add serialize function to Cluster (public)
- Add serialize function cluste descriptor (not public)
- Add flag to all Cluster constructors to create mock chips
- Add a test

### Testing
CI + new test

### API Changes

Added create_mock_chips to Cluster constuctors, but changes not needed in UMD clients.

cfjchu commented on device/tt_cluster_descriptor.cpp:
```diff
@@ -896,3 +903,75 @@ tt_ClusterDescriptor::get_chips_grouped_by_closest_mmio() const {
+    }
+    out << YAML::EndMap;
+
+    out << YAML::EndMap;
+
+    // Output YAML to file
+    std::filesystem::path temp_path = std::filesystem::temp_directory_path();
+    std::string cluster_path_dir_template = temp_path / "umd_XXXXXX";
+    std::filesystem::path cluster_path_dir = mkdtemp(cluster_path_dir_template.data());
+    std::filesystem::path cluster_path = cluster_path_dir / "cluster_descriptor.yaml";
+    std::ofstream file(cluster_path);
+    file << out.c_str();
+    file.close();
+
+    return cluster_path;
+}
```
can we separate out serialization from file I/O? 

Something like:

```
std::string tt_ClusterDescriptor::serialize() const {
    YAML::Emitter out;
    ...
    return out.c_str();
}

std::filesystem::path tt_ClusterDescriptor::serialize_to_file() const {
    auto serialized = this->serialize();
    ...
    // write to file and return path

}
```

pjanevskiTT replied:
done

cfjchu commented on tests/api/test_cluster.cpp:
```diff
@@ -279,3 +272,10 @@ TEST(ClusterAPI, DynamicTLB_RW) {
+TEST(ClusterAPI, TestClusterSerialize) {
+    std::string cluster_path = tt::umd::Cluster::serialize();
+    std::unordered_map<chip_id_t, HarvestingMasks> simulated_harvesting_masks = {};
+    std::unique_ptr<Cluster> cluster = std::make_unique<Cluster>(
+        tt_ClusterDescriptor::create_from_yaml(cluster_path), 1, false, false, true, simulated_harvesting_masks, true);
```
@pjanevskiTT Thanks! For the creation from yaml path, is the important flag here the `create_mock_chips` boolean? Do you have more information on the mock chips? 

This path:

```
    std::unique_ptr<Cluster> cluster = std::make_unique<Cluster>(
        tt_ClusterDescriptor::create_from_yaml(cluster_path), 1, false, false, true, simulated_harvesting_masks, true);
```
can be used on a machine without a silicon device?

pjanevskiTT replied:
The idea behind adding create_mock_chips is that each constructor can be called such that you get mock chips. Mock chips should probably have all the IO functions that interact with device empty. But you can still read all the relevant info of the chip. The things is that all these constructors apart from the one I added need the device to create cluster descriptor.
The call you pasted in the comment is a way to use Cluster on a machine without silicon device

pjanevskiTT replied:
I'll resolve the comment, feel free to write more if you have more questions

pjanevskiTT replied:
I figured that skip_driver_allocs was actually same case as create_mock_chips (don't touch silicon). Renamed that var to create_mock_chips.

Basically your call can now be (note there is no last argument). 

```cpp
    std::unique_ptr<Cluster> cluster = std::make_unique<Cluster>(
        tt_ClusterDescriptor::create_from_yaml(cluster_path), 1, true, false, true, simulated_harvesting_masks);
```

You can take a look at the test as well

broskoTT commented on device/api/umd/device/cluster.h:
```diff
@@ -758,34 +760,17 @@ class Cluster : public tt_device {
-     * @param clean_system_resource Specifies if host state from previous runs needs to be cleaned up.
-     * @param perform_harvesting Allow the driver to modify the SOC descriptors per chip.
-     * @param simulated_harvesting_masks
-     */
```
Leave some comment on the function

broskoTT commented on tests/api/test_cluster.cpp:
```diff
@@ -70,13 +70,6 @@ TEST(ApiClusterTest, DifferentConstructors) {
-    // to harvesting info.
-    // // 4. Constructor for creating a cluster with mock chip.
-    // umd_cluster = Cluster::create_mock_cluster();
-    // umd_cluster = nullptr;
```
Yeah can you add another test which demonstrates creating Cluster with mock chips?

broskoTT replied:
This is a general comment. I want to make sure we've covered all the scenarios which we want to cover. Maybe that deserves a separate api test (remember, api tests are not a place for testing code, rather for showcasing the usage):

- Creating a cluster, don' care about cluster descriptor
- Forging a custom cluster descriptor and creating a cluster using it
- Forging a custom cluster descriptor and creating a mock cluster using it
- Create a cluster and serialize cluster descriptor yaml so it can be used later, after modification or not, in one of the previous two scenarios.

@cfjchu are we missing some scenario that you folks want to have?

cfjchu replied:
Yes, I think these scenarios capture it. I would also like the ability to extend `Chip` or `MockChip` and provide implementations for intercepting the interactions that would've gone to an actual chip. 

For example, it would be useful  to have an ability to create a mock 64-chip cluster in an offline context where interactions to first 32-chips get transmitted as RPCs to one host galaxy and interactions with the other 32-chips get transmitted as RPCs to the other galaxy. Something similar to what is already being done with the simulator.

cfjchu replied:
and big +1 for adding more tests that exercise this user-flow

broskoTT replied:
@pjanevskiTT sounds like we might benefit from the constructor accepting vector of custom made Chips :) let's put more thought into it in the near future

broskoTT commented on device/api/umd/device/types/cluster_descriptor_types.h:
```diff
@@ -51,6 +52,33 @@ enum BoardType : uint32_t {
     UNKNOWN,
 };
 
+inline std::string board_type_to_string(const BoardType board_type) {
```
it's fine, ignore

pjanevskiTT commented:
@broskoTT fyi, `skip_driver_allocs` had the same purpose as `create_mock_chips`. I renamed it and fixed few bugs which we had when creating mock chips
